taskcat/amiupdater.py

```diff
@@ -0,0 +1,369 @@
+        # For reach RegionalCodename that we've generated....
+        pool.map(cls._per_rcn_ami_fetch, RegionalCodename.objects())
+
+    @classmethod
+    def _per_rcn_ami_fetch(cls, rcn):
+        rcn._results = AMIUpdater.client_factory.get('ec2', rcn.region).describe_images(Filters=rcn.filters)['Images']
+
+    @classmethod
+    def parse_api_results(cls):
+        raw_ami_names = {}
+        region_codename_result_list = []
+
+        # For each RegionalCodename.
+        #     Create a Dictionary like so:
+        #     CODENAME
+        #       - REGION_NAME
+        #           [{RAW_API_RESULTS_1}, {RAW_API_RESULTS_2}, {RAW_API_RESULTS_N}]
+        for rcn in RegionalCodename.objects():
+            if rcn.cn in raw_ami_names.keys():
+                raw_ami_names[rcn.cn][rcn.region] = [APIResultsData(rcn.cn, x['ImageId'], int(datetime.datetime.strptime(x['CreationDate'], "%Y-%m-%dT%H:%M:%S.%fZ").timestamp()), rcn.region) for x in rcn._results]
+            else:
+                raw_ami_names.update({rcn.cn: {rcn.region: [APIResultsData(rcn.cn, x['ImageId'], int(datetime.datetime.strptime(x['CreationDate'], "%Y-%m-%dT%H:%M:%S.%fZ").timestamp()), rcn.region) for x in rcn._results]}})
+
+        APIResultsData.custom_comparisons = True
+        for codename, regions in raw_ami_names.items():
+            for region, results_list in regions.items():
+                latest_ami = sorted(results_list, reverse=True)[0]
+                region_codename_result_list.append(latest_ami)
+        APIResultsData.custom_comparisons = False
+
+        APIResultsData.results = region_codename_result_list
+
+
+class RegionalCodename(Codenames):
+    _x = {}
+
+    def __new__(cls, *args, **kwargs):
+        # A word on this.
+        # - An instance of RegionalCodename is a representation of CODENAME and AMINAME.
+        # Since this module allows for multiple templates, *and* interrogates all templates at once,
+        # there's a risk I could end up with multiple instantations for each CODENAME/AMINAME combination.
+        # I maintain a dictionary of CODENAMEREGION -> Instance mappings, so if this class is instantated with the same
+        # arguments twice, the exact same memory pointer is returned.
+        try:
+            region = kwargs.get('region')
+            cn = kwargs.get('cn')
+        except KeyError:
+            raise
+        k = "{}{}".format(cn,region)
+        if k in cls._x.keys():
+            instance = cls._x.get(k)
+        else:
+            instance = super(Codenames, cls).__new__(cls)
+            cls._x[k] = instance
+        return instance
+
+    @classmethod
+    def objects(cls):
+        return [z for z in cls._x.values() if z.filters]
+
+    def __init__(self, cn, region, filters=None, *args, **kwargs):
+        self.region = region
+        self._results = None
+        self._filters = filters
+        super(RegionalCodename, self).__init__(cn, region, *args, **kwargs)
+
+class TemplateClass(object):
```

Comment:
looks like only regions method needs to be a classmethod, others can all be static.